### PR TITLE
Adding type to native markup for Prebid mobile

### PR DESCRIPTION
**Pull request by jaiminpanchal27**
This PR implements https://github.com/prebid/prebid-server/issues/1041

### CODE REVIEW

File: exchange/bidder.go [hunk 1]
```diff
@@ -178,6 +225,26 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+	}
+	return nil, errors.New("Could not find native imp")
+}
+
+func getAssetByID(id int64, assets []nativeRequests.Asset) nativeRequests.Asset {
+	var asset nativeRequests.Asset
+	for i := 0; i < len(assets); i++ {
+		if id == assets[i].ID {
+			asset = assets[i]
+			break
```
Comment (SyntaxNode) [inline exchange/bidder.go]: Consider returning asset here and returning nil at the end.

File: exchange/bidder.go [hunk 2]
```diff
@@ -178,6 +225,26 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+		if impInRequest.ID == impID && impInRequest.Native != nil {
+			return impInRequest.Native, nil
+		}
+	}
+	return nil, errors.New("Could not find native imp")
+}
+
+func getAssetByID(id int64, assets []nativeRequests.Asset) nativeRequests.Asset {
+	var asset nativeRequests.Asset
+	for i := 0; i < len(assets); i++ {
```
Comment (SyntaxNode) [inline exchange/bidder.go]: Consider using the `range` syntax of `for`.

File: exchange/bidder.go [hunk 3]
```diff
@@ -153,6 +156,50 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+							nativeImp, err := getNativeImpByImpID(bidResponse.Bids[i].Bid.ImpID, request)
+							if err != nil {
+								errs = append(errs, err)
+							}
+							var nativePayload nativeRequests.Request
+							if err := json.Unmarshal(json.RawMessage((*nativeImp).Request), &nativePayload); err != nil {
+								errs = append(errs, err)
+							}
+
+							for j := 0; j < len(nativeMarkup.Assets); j++ {
```
[the remark this answers is not in the document]
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Done

File: exchange/bidder.go (hunk 2 above)
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Done

File: exchange/bidder.go (hunk 1 above)
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Done

File: exchange/bidder.go [hunk 4]
```diff
@@ -153,6 +156,50 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
 					}
 				}
 
+				// Only do this for request from mobile app
+				if request.App != nil {
+					for i := 0; i < len(bidResponse.Bids); i++ {
+						if bidResponse.Bids[i].BidType == openrtb_ext.BidTypeNative {
+							var nativeMarkup nativeResponse.Response
+							if err := json.Unmarshal(json.RawMessage(bidResponse.Bids[i].Bid.AdM), &nativeMarkup); err != nil {
```
[the remark this answers is not in the document]
Comment (jaiminpanchal27) [reply exchange/bidder.go]: json.Unmarshal expects a slice of byte. Here AdM is of type string and hence I used json.RawMessage

File: exchange/bidder.go [hunk 5]
```diff
@@ -153,6 +156,50 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+								// Some bidders are not following IAB spec when doing native. One such example is Facebook. We are not ignoring such bids
```
[the remark this answers is not in the document]
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Updated comment. I hope it makes more sense now

File: exchange/bidder.go [hunk 6]
```diff
@@ -153,6 +156,50 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
 					}
 				}
 
+				// Only do this for request from mobile app
+				if request.App != nil {
+					for i := 0; i < len(bidResponse.Bids); i++ {
```
[the remark this answers is not in the document]
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Done

**Comment by jaiminpanchal27** [comment]
@SyntaxNode Thanks for feedback. I pushed an update

File: exchange/bidder.go [hunk 7]
```diff
@@ -178,6 +199,64 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
 	return seatBid, errs
 }
 
+func addNativeTypes(bid *openrtb.Bid, request *openrtb.BidRequest) (*nativeResponse.Response, []error) {
+	var errs []error
+	var nativeMarkup *nativeResponse.Response
+	if err := json.Unmarshal(json.RawMessage(bid.AdM), &nativeMarkup); err != nil {
```
Comment (SyntaxNode) [inline exchange/bidder.go]: Why is it necessary to cast the message to `json.RawMessage` here and below on line 215?

Comment (SyntaxNode) [reply exchange/bidder.go]: ... I didn't see your answer above. Doh!

File: exchange/bidder.go [hunk 8]
```diff
@@ -153,6 +156,24 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+						if bidResponse.Bids[i].BidType == openrtb_ext.BidTypeNative {
+							nativeMarkup, moreErrs := addNativeTypes(bidResponse.Bids[i].Bid, request)
+							errs = append(errs, moreErrs...)
+
+							if nativeMarkup != nil {
+								markup, err := json.Marshal(*nativeMarkup)
+								if err != nil {
+									errs = append(errs, err)
+								}
+								bidResponse.Bids[i].Bid.AdM = string(markup)
```
Comment (hhhjort) [inline exchange/bidder.go]: If there were errors in the marshalling, do we still want to we still want to try overwriting the AdM with whatever `json.Marshal()` managed to spit out?

**Comment by jaiminpanchal27** [comment]
@anwzhang Correct. If `request.App` flag is only set when the request is from mobile device, we are all set

File: exchange/bidder.go (hunk 8 above)
Comment (SyntaxNode) [reply exchange/bidder.go]: Good point. `markup` very well may be `nil`, of which `string(markup)` would succeed and result in an empty string.

File: exchange/bidder.go [hunk 9]
```diff
@@ -178,6 +199,64 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.Bi
...
+		// Some bidders are returning non-IAB complaiant native markup. In this case Prebid server will not be able to add types. E.g Facebook
+		return nil, errs
+	}
+
+	nativeImp, err := getNativeImpByImpID(bid.ImpID, request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	var nativePayload nativeRequests.Request
+	if err := json.Unmarshal(json.RawMessage((*nativeImp).Request), &nativePayload); err != nil {
```
Comment (hhhjort) [inline exchange/bidder.go]: We have a potential panic here. If `getNativeImpByImpID()` fails to find the imp, then `nativeImp` is nil. Trying to access `nil.Request` will then panic, as you cannot dereference a `nil`. You should probably abort above if `nativeImp` is `nil`. I don't think we need to abort here on `Unmarshal` errors as the worst that should happen is we fail asset lookups and don't do anything.

File: exchange/exchangetest/mobile-native-types.json [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      "mockResponse": {
+        "pbsSeatBid": {
+          "pbsBids": [
+            {
+              "ortbBid": {
+                "id": "apn-bid",
+                "impid": "my-imp-id",
+                "price": 10,
+                "adid": "1234",
+                "adm": "{\"assets\":[{\"id\":1,\"img\":{\"type\":3,\"url\":\"http://vcdn.adnxs.com/p/creative-image/f8/7f/0f/13/f87f0f13-230c-4f05-8087-db9216e393de.jpg\",\"w\":989,\"h\":742,\"ext\":{\"appnexus\":{\"prevent_crop\":0}}}},{\"title\":{\"text\":\"This is a Prebid Native Creative\"}},{\"id\":2,\"data\":{\"type\":1,\"value\":\"Prebid.org\"}},{\"id\":3,\"data\":{\"type\":2,\"value\":\"This is a Prebid Native Creative.  There are many like it, but this one is mine.\"}}],\"link\":{\"url\":\"http://some-url.com\"},\"imptrackers\":[\"http://someimptracker.com\"],\"jstracker\":\"some-js-tracker\"}",
```
Comment (hhhjort) [inline exchange/exchangetest/mobile-native-types.json]: In the mock response, don't you want the adapter to return native markup without types being set, so you can verify the code does actually set the types?

File: exchange/bidder.go (hunk 8 above)
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Agreed. Updated

File: exchange/bidder.go (hunk 9 above)
Comment (jaiminpanchal27) [reply exchange/bidder.go]: Done

File: exchange/exchangetest/mobile-native-types.json (hunk 10 above)
Comment (jaiminpanchal27) [reply exchange/exchangetest/mobile-native-types.json]: I misunderstood the exchange_test.go Now I moved my test to bidder_test.go

**Comment by jaiminpanchal27** [comment]
@hhhjort Pushed an update. Addressed all your feedback

**Pull request merged**